File: merge_sort/merge_sort.py
# @Author alacic gu


def merge_sort(arr):
    arr_size = len(arr)
    if arr_size > 1:
        mid = arr_size // 2

        L = arr[:mid]
        R = arr[mid:]

        merge_sort(L)
        merge_sort(R)

        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        while i < len(L):
            arr[k] = L[i]
            k += 1
            i += 1

        while j < len(R):
            arr[k] = R[j]
            k += 1
            j += 1


if __name__ == "__main__":
    arr = [1, 5, 6, 3, 4, 7, 8, 9, 5, 4, 7, 5]
    print(f"Before sorted, arr is {arr}")
    merge_sort(arr)
    print(f"After sorted, arr is {arr}")

File: merge_sort/mergeSort.cpp
/**
 * merge sort
 * @Author alacic gu
 */

#include <iostream>

using namespace std;

void mergeSort(int arr[], int arr_size);
void printArray(int arr[], int arr_size);

int main()
{
    int arr[]{1, 5, 9, 7, 8, 12, 5, 5, 5, 2};
    cout << "Before sorted, arr is ";
    int n = sizeof(arr) / sizeof(arr[0]);
    printArray(arr, n);
    mergeSort(arr, n);
    cout << "After sorted, arr is ";
    printArray(arr, n);
}

void mergeSort(int arr[], int arr_size)
{
    if (arr_size <= 1)
        return;

    int mid = arr_size / 2;
    auto *L = new int[mid];
    auto *R = new int[arr_size - mid];

    // copy value
    int i = 0, j = 0, k = 0;
    for (k = 0; k < arr_size; k++)
    {
        if (k < mid)
        {
            L[i] = arr[k];
            i++;
        }
        else
        {
            R[j] = arr[k];
            j++;
        }
    }

    // recursion
    mergeSort(L, mid);
    mergeSort(R, arr_size - mid);

    // merge step
    i = 0;
    j = 0;
    k = 0;
    while (i < mid && j < arr_size - mid)
    {
        if (L[i] <= R[j])
        {
            arr[k] = L[i];
            i++;
            k++;
        }
        else
        {
            arr[k] = R[j];
            j++;
            k++;
        }
    }

    while (i < mid)
    {
        arr[k] = L[i];
        i++;
        k++;
    }

    while (j < arr_size - mid)
    {
        arr[k] = R[j];
        j++;
        k++;
    }
    delete[] R;
    delete[] L;
}

void printArray(int arr[], int arr_size)
{
    for (int i = 0; i < arr_size - 1; i++)
    {
        cout << arr[i] << ", ";
    }
    cout << arr[arr_size - 1] << endl;
}
